release-lock-file-spec.md

```diff
@@ -1,74 +1,196 @@
+1. Release lockfiles are intended to be generated programatically by package management software as part of the release process.
+2. Release lockfiles will be consumed by package managers during tasks like installing package dependencies or building and deploying new releases.
+3. Release lockfiles will typically **not** be stored alongside the source, but rather by package registries *or* referenced by package registries and stored in something akin to IPFS.
 
 
-## Format
+## Keywords
 
-The canonical format for the package manifest is a JSON document containing a
-single JSON object.  
+### RFC2119
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in RFC 2119.
+
+* https://www.ietf.org/rfc/rfc2119.txt
+
+### Custom
+
+#### Prefixed vs Unprefixed
+
+A prefixed hexidecimal value begins with `'0x'`.  Unprefixed values have no
+prefix.  Unless otherwise specified, all hexidecimal values should be
+represented with the `'0x'` prefix.
+
+* Prefixed: `0xdeadbeef`
+* Unprefixed: `deadbeef`
+
+
+#### Bytecode
+
+The set of EVM instructions as produced by a compiler.  Unless otherwise
+specified this should be assumed to be hexidecimal encoded and prefixed with a
+`'0x'`.
+
+#### Unlinked Bytecode
+
+Unlinked bytecode is the hexidecimal representation of a contract's EVM
+instructions which contains placeholders which are referred to as *link
+references*.
+
+* Unlinked Bytecode: `606060405260e0600073__MathLib_______________________________634d536f`
+
+
+#### Linked Bytecode
+
+Linked bytecode is the hexidecimal representation of a contract's EVM instructions which has hadd all *link references* replaced with the desired *link values*
+
+* linked Bytecode: `606060405260e06000736fe36000604051602001526040518160e060020a634d536f`
+
+
+#### Link Reference
+
+A placeholder within the hexidecimal representation of a contract's EVM instructions.
+
+`606060405260e0600073__MathLib_______________________________634d536f` the
+
+In the bytecode
+`606060405260e0600073__MathLib_______________________________634d536f`, the
+substring `__MathLib_______________________________` is a link reference.
+
+
+#### Link Value
 
+A link value is the value which can be inserted in place of a link reference.
 
-## Filename
 
-When creating a release, convention is to name this document `<version>.json`.
+#### Linking
 
-* The `1.0.0` Release would be named `1.0.0.json`
+The act of replacing *link references* within some bytecode with *link values*.
 
-Package managers may keep their own copy of this file as part of their package installation process.  In these cases it is up to the package manager to decide on a naming convention.  Package managers **should** keep the `.json` file extension.
+
+#### Contract Type
+
+This term is used to refer to a specific contract or library in the package
+source.
```

Comment:
Referring a contract as a `library`, `abstract contract`, `interface`, etc. is an unfortunate overloading of the word "type". Both usages are correct here, but they describe different things. For the purposes of this document we want to describe the _name_ of the object that can be placed on the blockchain, and not necessarily the _type_ of object that is (i.e., `contract`, `abstract contract` or `library`). Some examples:

With the following code, the type is `A`:

```
contract A {
  public uint value;
  function A() { 
    value = 2;
  }
}
```

This code also has a type `A`: 

```
library A {
  function add(uint a, uint b) returns (uint) { return a + b; }
}
```

And so does this code: 

```
contract A {
  function add() returns (uint);
}
```

So it's not the _type_ of object, but rather than the _name_ of the object. We're using the word "type" here because instances have a type, in that we can say an instance of contract `A` is of type `A`.